clarity/src/vm/ast/parser_v2/mod.rs

```diff
@@ -0,0 +1,2518 @@
+        assert_eq!(
+            diagnostics[0].spans[0],
+            Span {
+                start_line: 1,
+                start_column: 36,
+                end_line: 1,
+                end_column: 38
+            }
+        );
+
+        let (stmts, diagnostics, success) = parse_collect_diagnostics(".foo.bar");
+        assert_eq!(success, true);
+        assert_eq!(stmts.len(), 1);
+        match &stmts[0].pre_expr {
+            PreSymbolicExpressionType::SugaredFieldIdentifier(contract_name, trait_name) => {
+                assert_eq!(contract_name.as_str(), "foo");
+                assert_eq!(trait_name.as_str(), "bar");
+            }
+            _ => panic!("failed to parse sugared trait identifier"),
+        }
+        assert_eq!(
+            stmts[0].span,
+            Span {
+                start_line: 1,
+                start_column: 1,
+                end_line: 1,
+                end_column: 8
+            }
+        );
+        assert_eq!(diagnostics.len(), 0);
+
+        let (stmts, diagnostics, success) = parse_collect_diagnostics(".foo.123");
+        assert_eq!(success, false);
+        assert_eq!(stmts.len(), 1);
+        match &stmts[0].pre_expr {
+            PreSymbolicExpressionType::SugaredFieldIdentifier(contract_name, trait_name) => {
+                assert_eq!(contract_name.as_str(), "foo");
+                assert_eq!(trait_name.as_str(), "placeholder");
+            }
+            _ => panic!("failed to parse sugared trait identifier"),
+        }
+        assert_eq!(diagnostics.len(), 1);
+        assert_eq!(diagnostics[0].message, "expected trait identifier");
+        assert_eq!(
+            diagnostics[0].spans[0],
+            Span {
+                start_line: 1,
+                start_column: 6,
+                end_line: 1,
+                end_column: 8
+            }
+        );
+
+        let (stmts, diagnostics, success) = parse_collect_diagnostics(".foo.veryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryverylong");
+        assert_eq!(success, false);
+        assert_eq!(stmts.len(), 1);
+        match &stmts[0].pre_expr {
+            PreSymbolicExpressionType::SugaredFieldIdentifier(contract_name, trait_name) => {
+                assert_eq!(contract_name.as_str(), "foo");
+                assert_eq!(trait_name.as_str(), "placeholder");
+            }
+            _ => panic!("failed to parse sugared trait identifier"),
+        }
+        assert_eq!(diagnostics.len(), 1);
+        assert_eq!(diagnostics[0].message, "illegal name (too long), 'veryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryverylong'");
+        assert_eq!(
+            diagnostics[0].spans[0],
+            Span {
+                start_line: 1,
+                start_column: 6,
+                end_line: 1,
+                end_column: 137
+            }
+        );
```

Comment:
Can you add a test for a principal with invalid Crockford 32 characters?

Reply:
Added in 621f4db9d